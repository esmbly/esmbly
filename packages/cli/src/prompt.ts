import inquirer from 'inquirer';
import { Config } from '@esmbly/types';
import {
  getAvailableOutputFormats,
  getAvailableTransformers,
} from '@esmbly/utils';

export async function promptForConfig(): Promise<Config> {
  const { input } = await inquirer.prompt([
    {
      message: 'Which files do you want to transform?',
      name: 'input',
      type: 'input',
    },
  ]);

  const { transformers } = await inquirer.prompt([
    {
      choices: await getAvailableTransformers(),
      message: 'Which transformer do you want to use?',
      name: 'transformers',
      type: 'checkbox',
    },
  ]);

  // TODO: Ask to install transformers here?

  const { output } = await inquirer.prompt([
    {
      choices: await getAvailableOutputFormats(transformers),
      message: 'Which output formats to you want to use?',
      name: 'output',
      type: 'checkbox',
    },
  ]);

  return { input: [input], output, transformers };
}
