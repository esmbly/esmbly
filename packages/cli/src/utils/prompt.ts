import inquirer from 'inquirer';
import {
  TransformerInfo,
  getAvailableOutputFormats,
  getAvailableTransformers,
} from '@esmbly/utils';
import stringify from 'stringify-object';
import chalk from 'chalk';

interface PromptedConfig {
  module: string;
  install: boolean;
  transformers: TransformerInfo[];
  pkgManager: string;
}

function createRequireString(transformers: TransformerInfo[]): string {
  return transformers
    .map((t: TransformerInfo) => `const ${t.const} = require('${t.pkg}')`)
    .join('\n');
}

function createConfigString(
  input: string,
  output: string[],
  outDir: string,
): string {
  return stringify({
    input: [input],
    transformers: '<TRANSFORMERS>',
    // eslint-disable-next-line
    output: output.map((format: string) => ({ format, outDir })),
  });
}

function createTransformerCalls(transformers: TransformerInfo[]): string {
  /**
   * Produces:
   * [
   *   Flow.createTransformer(),
   *   Wasm.createTransformer()
   * ]
   */
  return `${transformers.reduce(
    (calls, t) => `${calls}\n    ${t.const}.createTransformer(),`,
    '[',
  )}\n  ]`;
}

function createModuleString(
  input: string,
  transformers: TransformerInfo[],
  output: string[],
  outDir: string,
): string {
  const requireString = createRequireString(transformers);
  const configString = createConfigString(input, output, outDir);
  const transformerCalls = createTransformerCalls(transformers);

  return `${requireString}\n\nmodule.exports = ${configString.replace(
    `'<TRANSFORMERS>'`,
    transformerCalls,
  )}\n`;
}

function isValidCombination(transformers: TransformerInfo[]): boolean {
  let isValid = true;

  for (let i = 0; i < transformers.length - 1; i += 1) {
    const current = transformers[i];
    const next = transformers[i + 1];

    if (!current.to.includes(next.from)) {
      isValid = false;
      break;
    }
  }

  return isValid;
}

export async function promptForConfig(): Promise<PromptedConfig> {
  const availableTransformers = getAvailableTransformers();

  const { input } = await inquirer.prompt([
    {
      message: 'Which files do you want to transform?',
      name: 'input',
      suffix: chalk.dim(' e.g ./**/*.js'),
      type: 'input',
    },
  ]);

  const { selectedTransformers } = await inquirer.prompt([
    {
      choices: availableTransformers.map(t => ({
        name: `${t.name}\t${chalk.dim(`${t.from} -> ${t.to.join(' | ')}`)}`,
        short: t.name,
        value: t.name,
      })),
      default: [],
      message: 'Which transformers do you want to use?',
      name: 'selectedTransformers',
      type: 'checkbox',
    },
  ]);

  const transformers = availableTransformers.filter(t =>
    selectedTransformers.includes(t.name),
  );

  if (!isValidCombination(transformers)) {
    const combination = transformers.reduce(
      (str: string, transformer: TransformerInfo, i: number) => {
        if (i === 0) {
          return `${transformer.name}`;
        }

        return `${str} -> ${transformer.name}`;
      },
      '',
    );
    throw new Error(
      `Transformations in the order ${combination} is not possible`,
    );
  }

  const { output } = await inquirer.prompt([
    {
      choices: getAvailableOutputFormats(transformers),
      message: 'Which formats to you want to output?',
      name: 'output',
      type: 'checkbox',
      when: transformers.length > 0,
    },
  ]);

  const { outDir } = await inquirer.prompt([
    {
      message: 'To which directory do you want to output files?',
      name: 'outDir',
      suffix: chalk.dim(' e.g ./dist'),
      type: 'input',
    },
  ]);

  const { install } = await inquirer.prompt([
    {
      default: true,
      message: 'Do you want to install the selected transformers and the CLI?',
      name: 'install',
      type: 'confirm',
      when: transformers.length > 0,
    },
  ]);

  const { pkgManager } = await inquirer.prompt([
    {
      choices: ['npm', 'yarn'],
      message: 'Which package manager do you want to use for installation?',
      name: 'pkgManager',
      type: 'list',
      when: install === true,
    },
  ]);

  return {
    install,
    module: createModuleString(input, transformers, output, outDir),
    pkgManager,
    transformers,
  };
}
