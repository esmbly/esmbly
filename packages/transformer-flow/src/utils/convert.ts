import {
  FlowType,
  TSType,
  nullLiteralTypeAnnotation,
  numberTypeAnnotation,
  stringTypeAnnotation,
  tsNullKeyword,
  tsNumberKeyword,
  tsStringKeyword,
} from '@babel/types';
import { Format } from '@esmbly/types';

interface TMap {
  [type: string]: {
    [key: string]: Function;
  };
}

const TypeMap: TMap = {
  null: {
    [Format.Flow]: nullLiteralTypeAnnotation,
    [Format.TypeScript]: tsNullKeyword,
  },
  number: {
    [Format.Flow]: numberTypeAnnotation,
    [Format.TypeScript]: tsNumberKeyword,
  },
  string: {
    [Format.Flow]: stringTypeAnnotation,
    [Format.TypeScript]: tsStringKeyword,
  },
};

// @ts-ignore
export function convertAnnotation(
  node: TSType | FlowType,
  to: string,
): TSType | FlowType | null {
  switch (node.type) {
    case 'TSStringKeyword':
    case 'StringTypeAnnotation':
      return TypeMap.string[to]();
    case 'TSNumberKeyword':
    case 'NumberTypeAnnotation':
      return TypeMap.number[to]();
    default:
      return null;
  }
}

export function toTs(node: TSType | FlowType): TSType {
  return convertAnnotation(node, 'TypeScript') as TSType;
}
