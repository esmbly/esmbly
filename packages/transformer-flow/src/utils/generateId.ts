const candidates = 'abcdefghijklmnopqrstuvwxyz'.split('');

export default (usedIdentifiers: string[]): string => {
  const identifier = candidates.find(
    (candidate: string) => usedIdentifiers.indexOf(candidate) < 0,
  );
  if (!identifier) {
    throw new Error('Could not generate a free identifier');
  }
  usedIdentifiers.push(identifier);
  return identifier;
};
