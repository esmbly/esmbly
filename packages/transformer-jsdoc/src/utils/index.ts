export * from './toKeyword';
export * from './toTypeAnnotation';
export * from './isParam';
export * from './isReturn';
