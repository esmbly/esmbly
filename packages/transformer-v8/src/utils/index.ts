export * from './toKeyword';
export * from './toTypeAnnotation';
export * from './Profiler';
