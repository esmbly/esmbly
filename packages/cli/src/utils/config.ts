import path from 'path';
import { Config, InitOptions } from '@esmbly/types';
import {
  exists,
  getRelativePathTo,
  getRoot,
  readFile,
  writeFile,
} from '@esmbly/utils';
import { promptForConfig } from './prompt';
import { installCLI, installTransformers } from './install';

export const DEFAULT_FILE = 'esmbly.config.js';

export async function getTemplateConfig(): Promise<string> {
  const templatePath = path.resolve(
    __dirname,
    `../../defaults/${DEFAULT_FILE}`,
  );
  const template = await readFile(templatePath);
  return template.toString();
}

export function getDefaultConfigPath(): string {
  const root = getRoot();
  return path.join(root, DEFAULT_FILE);
}

type ConfigFnA = () => Config;
type ConfigFnB = () => Config[];
type ConfigFn = ConfigFnA | ConfigFnB;

function toArrayConfig(maybeArrayConfig: Config | Config[]): Config[] {
  if (Array.isArray(maybeArrayConfig)) {
    return maybeArrayConfig;
  }

  return [maybeArrayConfig];
}

export async function readConfig(
  customPath?: string,
  requirer: (requirePath: string) => unknown = require,
): Promise<Config[]> {
  const configPath = customPath || (await getDefaultConfigPath());
  const config = requirer(configPath) as Config | Config[] | ConfigFn;

  if (typeof config === 'function') {
    const configFnResult = await config();
    return toArrayConfig(configFnResult);
  }

  return toArrayConfig(config);
}

export async function createConfig(
  options: InitOptions,
): Promise<{
  root: string;
  fileName: string;
}> {
  const root = getRoot();
  const defaultConfigPath = getDefaultConfigPath();

  if ((await exists(defaultConfigPath)) && !options.force) {
    const relativeConfigPath = getRelativePathTo(defaultConfigPath);
    throw new Error(
      `Config ${relativeConfigPath} already exists. Remove that file first.`,
    );
  } else if (options.default) {
    const template = await getTemplateConfig();
    await writeFile(defaultConfigPath, template, { overwrite: options.force });
  } else {
    const config = await promptForConfig();

    if (config.install) {
      await installTransformers(config.transformers, config.pkgManager);
      await installCLI(config.pkgManager);
    }

    await writeFile(defaultConfigPath, config.module, {
      overwrite: options.force,
    });
  }

  return { fileName: DEFAULT_FILE, root };
}
