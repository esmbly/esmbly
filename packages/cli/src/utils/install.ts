import { exec } from 'child_process';
import { promisify } from 'util';
import { TransformerInfo } from '@esmbly/utils';
import ora from 'ora';

export function getPkgManagerCommand(pkgManager: string): string {
  switch (pkgManager) {
    case 'npm':
      return 'npm install --save-dev';
    case 'yarn':
      return 'yarn add --dev';
    default:
      throw new Error(`Package manager ${pkgManager} is not supported`);
  }
}

export async function installTransformers(
  transformers: TransformerInfo[],
  pkgManager: string,
): Promise<void> {
  const command = getPkgManagerCommand(pkgManager);

  for (const transformer of transformers) {
    const spinner = ora(`Installing: ${transformer.pkg}`).start();
    await promisify(exec)(`${command} ${transformer.pkg}`);
    spinner.succeed(`Installed: ${transformer.pkg}`);
  }
}

export async function installCLI(pkgManager: string): Promise<void> {
  const command = getPkgManagerCommand(pkgManager);

  const spinner = ora('Installing: @esmbly/cli').start();
  await promisify(exec)(`${command} @esmbly/cli`);
  spinner.succeed('Installed: @esmbly/core');
}
