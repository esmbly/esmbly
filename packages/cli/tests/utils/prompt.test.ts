import inquirer from 'inquirer';
import { promptForConfig } from '../../src/utils/prompt';

describe('promptForConfig', () => {
  it('asks the correct questions', async () => {
    const promptSpy = jest.spyOn(inquirer, 'prompt');
    // @ts-ignore
    promptSpy.mockImplementation(([{ name }]) => {
      switch (name) {
        case 'input':
          return Promise.resolve({ input: './src/index.js' });
        case 'selectedTransformers':
          return Promise.resolve({
            selectedTransformers: ['transformer-flow', 'transformer-wasm'],
          });
        case 'output':
          return Promise.resolve({ output: ['TypeScript', 'WebAssembly'] });
        case 'outDir':
          return Promise.resolve({ outDir: 'dist' });
        case 'install':
          return Promise.resolve({ install: true });
        case 'pkgManager':
          return Promise.resolve({ pkgManager: 'npm' });
        default:
          throw new Error(`asked for: ${name}`);
      }
    });
    const config = await promptForConfig();
    expect(promptSpy.mock.calls).toMatchSnapshot();
    expect(config).toMatchSnapshot();
  });

  it('throws an error on invalid transformer combinations', async () => {
    const promptSpy = jest.spyOn(inquirer, 'prompt');
    // @ts-ignore
    promptSpy.mockImplementation(([{ name }]) => {
      switch (name) {
        case 'input':
          return Promise.resolve({ input: './src/index.js' });
        case 'selectedTransformers':
          return Promise.resolve({
            selectedTransformers: ['transformer-flow', 'transformer-jsdoc'],
          });
        default:
          throw new Error(`asked for: ${name}`);
      }
    });
    await expect(promptForConfig()).rejects.toMatchSnapshot();
  });
});
