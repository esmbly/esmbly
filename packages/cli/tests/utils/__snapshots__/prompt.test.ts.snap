// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`promptForConfig asks the correct questions 1`] = `
Array [
  Array [
    Array [
      Object {
        "message": "Which files do you want to transform?",
        "name": "input",
        "suffix": "[2m e.g ./**/*.js[22m",
        "type": "input",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "transformer-flow	[2mFlow -> TypeScript[22m",
            "short": "transformer-flow",
            "value": "transformer-flow",
          },
          Object {
            "name": "transformer-jsdoc	[2mJSDoc -> TypeScript[22m",
            "short": "transformer-jsdoc",
            "value": "transformer-jsdoc",
          },
          Object {
            "name": "transformer-v8	[2mV8 -> TypeScript[22m",
            "short": "transformer-v8",
            "value": "transformer-v8",
          },
          Object {
            "name": "transformer-wasm	[2mTypeScript -> WebAssembly | Wat | Asm | AssemblyScript[22m",
            "short": "transformer-wasm",
            "value": "transformer-wasm",
          },
        ],
        "default": Array [],
        "message": "Which transformers do you want to use?",
        "name": "selectedTransformers",
        "type": "checkbox",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "TypeScript",
          "WebAssembly",
          "Wat",
          "Asm",
          "AssemblyScript",
        ],
        "message": "Which formats to you want to output?",
        "name": "output",
        "type": "checkbox",
        "when": true,
      },
    ],
  ],
  Array [
    Array [
      Object {
        "message": "To which directory do you want to output files?",
        "name": "outDir",
        "suffix": "[2m e.g ./dist[22m",
        "type": "input",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "default": true,
        "message": "Do you want to install the selected transformers and the CLI?",
        "name": "install",
        "type": "confirm",
        "when": true,
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "npm",
          "yarn",
        ],
        "message": "Which package manager do you want to use for installation?",
        "name": "pkgManager",
        "type": "list",
        "when": true,
      },
    ],
  ],
]
`;

exports[`promptForConfig asks the correct questions 2`] = `
Object {
  "install": true,
  "module": "const Flow = require('@esmbly/transformer-flow')
const Wasm = require('@esmbly/transformer-wasm')

module.exports = {
	input: [
		'./src/index.js'
	],
	transformers: [
    Flow.createTransformer(),
    Wasm.createTransformer(),
  ],
	output: [
		{
			format: 'TypeScript',
			outDir: 'dist'
		},
		{
			format: 'WebAssembly',
			outDir: 'dist'
		}
	]
}
",
  "pkgManager": "npm",
  "transformers": Array [
    Object {
      "const": "Flow",
      "from": "Flow",
      "name": "transformer-flow",
      "pkg": "@esmbly/transformer-flow",
      "to": Array [
        "TypeScript",
      ],
    },
    Object {
      "const": "Wasm",
      "from": "TypeScript",
      "name": "transformer-wasm",
      "pkg": "@esmbly/transformer-wasm",
      "to": Array [
        "WebAssembly",
        "Wat",
        "Asm",
        "AssemblyScript",
      ],
    },
  ],
}
`;

exports[`promptForConfig throws an error on invalid transformer combinations 1`] = `[Error: Transformations in the order transformer-flow -> transformer-jsdoc is not possible]`;
